snapcraft/internal/meta/slots.py

```diff
@@ -0,0 +1,195 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+import os
+import re
+
+from collections import OrderedDict
+from snapcraft.internal.meta.errors import SlotValidationError
+from typing import Any, Dict, List, Set  # noqa: F401
+
+logger = logging.getLogger(__name__)
+
+
+class Slot:
+    """Representation of a generic snap slot."""
+
+    def __init__(self, *, slot_name: str = "") -> None:
+        self._slot_dict = dict()  # type: Dict[str, Any]
+        self.slot_name = slot_name
+
+    def validate(self):
+        pass
+
+    @classmethod
+    def from_dict(cls, slot_dict: Dict[str, Any], slot_name: str = ""):
+        # If we explicitly support the type, use it instead.
+        if slot_dict.get("interface") == "content":
+            return ContentSlot.from_dict(slot_dict, slot_name=slot_name)
+        elif slot_dict.get("interface") == "dbus":
+            return DbusSlot.from_dict(slot_dict, slot_name=slot_name)
+
+        # Handle the general case.
+        slot = Slot(slot_name=slot_name)
+        slot._slot_dict = slot_dict
+        return slot
+
+    def to_yaml(self):
+        return OrderedDict(self._slot_dict)
+
+    def __repr__(self):
+        return repr(self.to_yaml())
+
+    def __str__(self):
+        return str(self.to_yaml())
+
+
+class ContentSlot(Slot):
+    """Representation of a snap content slot."""
+
+    def __init__(
+        self,
+        *,
+        slot_name: str = "",
+        read: List[str] = None,
+        write: List[str] = None,
+        use_source_key: bool = True
+    ) -> None:
+        super().__init__(slot_name=slot_name)
+
+        self._plug_dict = dict()  # type: Dict[str, Any]
+        self.interface = "content"  # type: str
+
+        if read is None:
+            self.read = list()  # type: List[str]
+        else:
+            self.read = read
+
+        if write is None:
+            self.write = list()  # type: List[str]
+        else:
+            self.write = write
+
+        self.use_source_key = use_source_key
+
+    def validate(self):
+        if self.interface != "content":
+            raise SlotValidationError(message="interface={}".format(self.interface))
+
+        if not self.read and not self.write:
+            raise SlotValidationError(message="`read` or `write` is required for slot")
+
+    def get_content_dirs(self, installed_path: str) -> Set[str]:
+        content_dirs = set()  # type: Set[str]
+
+        for path in self.read + self.write:
+            # Strip leading "$SNAP" and "/".
+            path = re.sub(r"^\$SNAP", "", path)
+            path = re.sub(r"^/", "", path)
+            content_dirs.add(os.path.join(installed_path, path))
```

Comment:
just for prettiness, do you want to also strip `./`?

Reply:
As in, the input path may list `./` like `./bin/...` ?

Reply:
Added ./ as well, but do correct me if I'm mistaking the intent!